### ISSUE TITLE
[BUG] Need to handle OSError EDEADLK/BlockingIOError on MacOS

### ISSUE THREAD

**Issue by ikus060** [rdiff-backup/rdiff-backup#1027]
## Bug summary

I'm facing a situation where the filesystem on MacOs is raising an OS Error 11 on read operation. It's most likely an issue with OneDrive on MacOS. Such a pain.

```
WARNING: Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'BlockingIOError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Sending back exception '[Errno 11] Resource deadlock avoided' of type <class 'BlockingIOError'> with traceback   File "rdiff_backup/connection.py", line 509, in _answer_request
  File "rdiff_backup/backup.py", line 197, in patch
  File "rdiff_backup/rorpiter.py", line 146, in __call__
  File "rdiff_backup/backup.py", line 600, in fast_process_file
  File "rdiff_backup/backup.py", line 655, in _patch_to_temp
  File "rdiff_backup/backup.py", line 702, in _patch_snapshot_to_temp
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Cleaning up
```

Reading more about this error code, I understand that we should better handle this error and effectively try again when this is raised.

## Version, Python, Operating System

rdiff-backup: 2.2.6
Python 3.
MacOS

## rdiff-backup call

Called trough Minarca. But it should not make a difference.

## What happened and what did you expect?

Honestly, I'm in unknown chart here. The combination of technology surprise me. Clearly OneDrive is buggy but it should not prevent rdiff-backup from completing the backup. At the moment when this error is raised, the backup completely stop.

I will do a quick fix and add `errno.EAGAIN` to the list of `_robust_errno_list` for testing.

## More information

N/A

**Comment by ikus060** [comment]
Reading more on the subject, we might need to handle all these error code the way way: EAGAIN and EWOULDBLOCK.

**Comment by ericzolf** [comment]
way way ⇒ same way? :smile: 
Anyway, quote from the errno man page:

> All the error names specified by POSIX.1 must have distinct values, with the exception of EAGAIN and EWOULDBLOCK, which may  be the same. 
> On Linux, these two have the same value on all architectures.

It might be the same on MacOS (aka FreeBSD).

That would be quite a change: we would need to neither ignore it or choke on it, but retry it. This is a whole new logic level. Is this really necessary? I mean, backup under Windows chokes also on OneDrive, my solution was to exclude the folder in my backup.

**Comment by ikus060** [comment]
I learn something new. Linux vs MacOS errno values are completely different.

EDEADLK == Errno 11  on MacOS, but EDEADLK on Linux.

```
WARNING: Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'BlockingIOError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Sending back exception '[Errno 11] Resource deadlock avoided' of type <class 'BlockingIOError'> with traceback   File "rdiff_backup/connection.py", line 509, in _answer_request
  File "rdiff_backup/backup.py", line 197, in patch
  File "rdiff_backup/rorpiter.py", line 146, in __call__
  File "rdiff_backup/backup.py", line 600, in fast_process_file
  File "rdiff_backup/backup.py", line 655, in _patch_to_temp
  File "rdiff_backup/backup.py", line 702, in _patch_snapshot_to_temp
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Cleaning up
```

So I want to capture "Errno 11", on MacOS.  On MacOs, `errno.EDEADLK == 11`.

errno.EDEADLK` is already part of the `_robust_errno_list`, but why does rdiff-backup terminate. If should be capture by `check_common_error`

**Comment by ericzolf** [comment]
Yeah, EDEADLK on Linux is 35, where EAGAIN is indeed 11. A bit more context would be required to understand what's going on. Could you repeat the error in debug mode, or at least provide some more lines of log?

**Comment by ikus060** [comment]
I manage to capture a bit more log on the client side. 

```
rdiff-backup 2.2.6
* Executing remote command /usr/bin/ssh -oBatchMode=yes -oPreferredAuthentications=publickey -p 6603 -oUserKnownHostsFile=/Users/marie-claude/Library/Preferences/Minarca/known_hosts -oIdentitiesOnly=yes -i /Users/marie-claude/Library/Preferences/Minarca/id_rsa minarca@minarca.net 'minarca/6.0.0b11.dev3+g72a1ccf rdiff-backup/2.2.6 (Darwin 21.6.0 x86_64)'
WARNING: this command line interface is deprecated and will disappear, start using the new one as described with '--new --help'.
WARNING: Server will be called with deprecated command line interface to guarantee compatibility. It might lead to a deprecation warning from newer rdiff-backup versions. Use '--api-version 201' (or higher) to avoid it.
* API version agreed to be 200 with command b"/usr/bin/ssh -oBatchMode=yes -oPreferredAuthentications=publickey -p 6603 -oUserKnownHostsFile=/Users/marie-claude/Library/Preferences/Minarca/known_hosts -oIdentitiesOnly=yes -i /Users/marie-claude/Library/Preferences/Minarca/id_rsa minarca@minarca.net 'minarca/6.0.0b11.dev3+g72a1ccf rdiff-backup/2.2.6 (Darwin 21.6.0 x86_64)'"
* Using repository 'iMacdeMieClaude'
NOTE: Found interrupted initial backup in data directory iMacdeMieClaude/rdiff-backup-data. Removing...
* Sending back exception '[Errno 39] Directory not empty: b'iMacdeMieClaude/rdiff-backup-data'' of type <class 'OSError'> with traceback   File "rdiff_backup/connection.py", line 509, in _answer_request

* Unable to import module (py)xattr. Extended attributes not supported on filesystem at path /
* Unable to import module posix1e from pylibacl package. POSIX ACLs not supported on filesystem at path /
* Unable to import win32security module. Windows ACLs not supported by filesystem at path /
* Could not determine if source directory / permits trailing spaces or periods in filenames because we can't find any files with trailing dot/period. It will be treated as permitting such files, but none will exist if it doesn't, so it doesn't really matter and is harmless
* -----------------------------------------------------------------
Detected abilities for source (read only) file system:
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 Off
  Case sensitivity                             Off
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
* POSIX ACLs not supported by filesystem at path iMacdeMieClaude/rdiff-backup-data/rdiff-backup.tmp.0 due to exception '[Errno 95] Operation not supported: 'iMacdeMieClaude/rdiff-backup-data/rdiff-backup.tmp.0''
* Unable to import win32security module. Windows ACLs not supported by filesystem at path iMacdeMieClaude/rdiff-backup-data/rdiff-backup.tmp.0
* -----------------------------------------------------------------
Detected abilities for destination (read/write) file system:
  Ownership changing                           Off
  Hard linking                                 On
  fsync() directories                          On
  Directory inc permissions                    On
  High-bit permissions                         On
  Symlink permissions                          Off
  Extended filenames                           On
  Windows reserved filenames                   Off
  Access control lists                         Off
  Extended attributes                          On
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
* Backup: escape_dos_devices = 0
* Backup: escape_trailing_spaces = 0
* Given repository doesn't need to be regressed
NOTE: Starting mirror from source path / to destination path iMacdeMieClaude
* Processing file .
....
* Processing file ...
* Processing file ...
....
* Processing file <some/path/to/a/file>.pdf
* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'OSError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1518, in read
  File "rdiff_backup/hash.py", line 43, in read
  File "rdiff_backup/rpath.py", line 1891, in read_timeout

* Processing file ...
* Processing file ..
.....
* Processing file <some/path/to/a/file>.pdf
WARNING: Exception '[Errno 11] Resource deadlock avoided' raised of class '<class 'BlockingIOError'>':
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Sending back exception '[Errno 11] Resource deadlock avoided' of type <class 'BlockingIOError'> with traceback   File "rdiff_backup/connection.py", line 509, in _answer_request
  File "rdiff_backup/backup.py", line 197, in patch
  File "rdiff_backup/rorpiter.py", line 146, in __call__
  File "rdiff_backup/backup.py", line 600, in fast_process_file
  File "rdiff_backup/backup.py", line 655, in _patch_to_temp
  File "rdiff_backup/backup.py", line 702, in _patch_snapshot_to_temp
  File "rdiff_backup/robust.py", line 94, in check_common_error
  File "rdiff_backup/rpath.py", line 1586, in copy
  File "rdiff_backup/rpath.py", line 1620, in copy_reg_file
  File "rdiff_backup/rpath.py", line 1152, in write_from_fileobj
  File "rdiff_backup/rpath.py", line 1537, in copyfileobj
  File "rdiff_backup/rpath.py", line 1514, in read
  File "rdiff_backup/iterfile.py", line 495, in read

* Cleaning up
* Cleaning up
Traceback (most recent call last):
  File "minarca_client/main.py", line 266, in _rdiff_backup
  File "rdiffbackup/run.py", line 108, in main_run
  File "rdiffbackup/actions/backup.py", line 159, in run
  File "rdiff_backup/backup.py", line 39, in mirror_compat200
  File "rdiff_backup/connection.py", line 579, in __call__
  File "rdiff_backup/connection.py", line 434, in reval
BlockingIOError: [Errno 11] Resource deadlock avoided
```

1. This version of the run include a patch with read_timeout. With or without the patch is fail the same way.
2. I find it strange that some `Resource deadlock avoided` are raised as `OSError` and the last one as `BlockingIOError`.
3. Could you confirm if the exception is raised and captured on the client side ? I'm confused because of the `_answer_request` in the stack trace.

**Comment by ikus060** [comment]
> Could you confirm if the exception is raised and captured on the client side ?

That seams to be the issue. I ran the backup with a small subset of the file for debugging with `-v 9`. I can't share the full log for security reasons. 

According to my finding, the exception is raised on the client side with error (11) with class BlockingIOError. But the exception is handled on the server side. Since the errno are different from platform to platform, the exception is not capture by `robust`. The exception cause the process to terminate.

I'm surprised this problem was never hit before.

Now, that I drill down the main issue. I wonder how to get it fixed for 2.2.6 and future release.

For some errno, the values are identical for all supported platform. So for those, we don't need to do anything. But for few other error, the value are different and the exception are not mapped to a specific class.

|errno|linux|macos|windows|Python Exception|
|---|---|---|---|---|
| errno.EPERM | 1 | 1 | 1 | |
| errno.ENOENT | 2 | 2 | 2 | |
| errno.EACCES | 13 | 13 | 13 | |
| errno.EBUSY | 16 | 16 | 16 | |
| errno.EEXIST | 17 | 17 | 17 | |
| errno.ENOTDIR | 20 | 20 | 20 | |
| errno.EILSEQ* | 84 | 92 | 42 | None |
| errno.EBADF | 9 | 9 | 9 | |
| errno.ENAMETOOLONG* | 36 | 63 | 38 | None |
| errno.EINTR | 4 | 4 | 4 | |
| errno.ESTALE* | 116 | 70 | 10070 | None |
| errno.ENOTEMPTY* | 399 | 66 | 41 | None |
| errno.EIO | 5 | 5 | 5 | |
| errno.ETXTBSY*| 26 | 26 | 139 | None |
| errno.ESRCH | 3 | 3 | 3 | |
| errno.EINVAL | 22 | 22 | 22 | |
| errno.EDEADLK* | 35 | 11 | 36 | BlockingIOError |
| errno.EOPNOTSUPP* | 95 | 102 | 10045 | None |
| errno.ETIMEDOUT* | 110 | 60 | 10060 | TimeoutError |
| errno.EXDEV | 18 | 18 | 18 | |

As a quick workaround, I patched the server side to capture the BlockingIOError and TimeoutError which is equivalent to OSError with `errno.EDEADLK` or `errno.ETIMEDOUT`

@ericzolf What are you proposal to get this fixed properly ?

**Comment by ericzolf** [comment]
Probably, the only correct way would be to capture all exceptions on the side where they're generated and encapsulate them there into an rdiff-backup OS agnostic exception. That could be a hell of a job, but the only solution I see.
